File: crates/api-types/src/diary.rs
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use crate::movies::{MovieDto, ReviewDto};

#[derive(Debug, Clone, Serialize, Deserialize, utoipa::ToSchema)]
pub struct LogReviewRequest {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub external_metadata_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub manual_title: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub manual_release_year: Option<u16>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub manual_director: Option<String>,
    pub rating: u8,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub comment: Option<String>,
    pub watched_at: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[schema(value_type = Option<String>)]
    pub watch_medium: Option<domain::value_objects::WatchMedium>,
}

#[derive(Debug, Clone, Serialize, Deserialize, utoipa::ToSchema)]
pub struct DiaryEntryDto {
    pub movie: MovieDto,
    pub review: ReviewDto,
}

#[derive(Debug, Clone, Deserialize, utoipa::IntoParams)]
#[into_params(parameter_in = Query)]
pub struct DiaryQueryParams {
    pub limit: Option<u32>,
    pub offset: Option<u32>,
    pub sort_by: Option<String>,
    pub movie_id: Option<Uuid>,
    pub user_id: Option<Uuid>,
}

#[derive(Debug, Clone, Deserialize, utoipa::IntoParams)]
#[into_params(parameter_in = Query)]
pub struct ActivityFeedQueryParams {
    pub limit: Option<u32>,
    pub offset: Option<u32>,
    pub sort_by: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, utoipa::ToSchema)]
pub struct FeedEntryDto {
    pub movie: MovieDto,
    pub review: ReviewDto,
    pub user_id: Uuid,
    pub user_display_name: String,
    pub is_federated: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub actor_url: Option<String>,
}

#[derive(Debug, Clone, Deserialize, utoipa::IntoParams)]
#[into_params(parameter_in = Query)]
pub struct ExportQueryParams {
    /// Output format: `csv` (default) or `json`
    #[serde(default = "default_export_format")]
    pub format: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, utoipa::ToSchema)]
pub struct EditReviewRequest {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub rating: Option<u8>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub comment: Option<Option<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub watched_at: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[schema(value_type = Option<Option<String>>)]
    pub watch_medium: Option<Option<domain::value_objects::WatchMedium>>,
}

fn default_export_format() -> String {
    "csv".to_string()
}

File: spa/src/features/diary.ts
import { z } from "zod"
import {
  useInfiniteQuery,
  useMutation,
  useQuery,
  useQueryClient,
} from "@tanstack/react-query"
import type { DiaryEntryDto, Paginated } from "@/lib/api/common"
import { diaryEntryDtoSchema, movieDtoSchema, paginatedSchema, reviewDtoSchema } from "@/lib/api/common"
import { del, get, patch, post } from "@/lib/api/client"

const PAGE_SIZE = 20

export const diaryQueryParamsSchema = z.object({
  limit: z.number().optional(),
  offset: z.number().optional(),
  sort_by: z.string().optional(),
  movie_id: z.string().uuid().optional(),
  user_id: z.string().uuid().optional(),
})
export type DiaryQueryParams = z.infer<typeof diaryQueryParamsSchema>

export const diaryResponseSchema = paginatedSchema(diaryEntryDtoSchema)
export type DiaryResponse = Paginated<DiaryEntryDto>

export const logReviewRequestSchema = z.object({
  external_metadata_id: z.string().optional(),
  manual_title: z.string().optional(),
  manual_release_year: z.number().optional(),
  manual_director: z.string().optional(),
  rating: z.number(),
  comment: z.string().optional(),
  watched_at: z.string(),
  watch_medium: z.string().optional(),
})
export type LogReviewRequest = z.infer<typeof logReviewRequestSchema>

export const editReviewRequestSchema = z.object({
  rating: z.number().optional(),
  comment: z.string().nullable().optional(),
  watched_at: z.string().optional(),
  watch_medium: z.string().nullable().optional(),
})
export type EditReviewRequest = z.infer<typeof editReviewRequestSchema>

export const feedEntryDtoSchema = z.object({
  movie: movieDtoSchema,
  review: reviewDtoSchema,
  user_id: z.string().uuid(),
  user_display_name: z.string(),
  is_federated: z.boolean(),
  actor_url: z.string().optional(),
})
export type FeedEntryDto = z.infer<typeof feedEntryDtoSchema>

export const activityFeedQueryParamsSchema = z.object({
  limit: z.number().optional(),
  offset: z.number().optional(),
  sort_by: z.string().optional(),
})
export type ActivityFeedQueryParams = z.infer<typeof activityFeedQueryParamsSchema>

export const activityFeedResponseSchema = paginatedSchema(feedEntryDtoSchema)
export type ActivityFeedResponse = Paginated<FeedEntryDto>

export const exportQueryParamsSchema = z.object({
  format: z.string().optional(),
})
export type ExportQueryParams = z.infer<typeof exportQueryParamsSchema>

function getDiary(params?: DiaryQueryParams) {
  return get<DiaryResponse>("/diary", params)
}

export function logReview(data: LogReviewRequest) {
  return post("/reviews", data)
}

export function editReview(id: string, data: EditReviewRequest) {
  return patch(`/reviews/${id}`, data)
}

function deleteReview(id: string) {
  return del(`/reviews/${id}`)
}

function getActivityFeed(params?: ActivityFeedQueryParams) {
  return get<ActivityFeedResponse>("/activity-feed", params)
}

export function exportDiary(params?: ExportQueryParams) {
  return get<Blob>("/diary/export", params)
}

export const diaryKeys = {
  all: ["diary"] as const,
  list: (params?: Partial<DiaryQueryParams>) => [...diaryKeys.all, "list", params] as const,
  infinite: (params?: Partial<DiaryQueryParams>) => [...diaryKeys.all, "infinite", params] as const,
  feed: (params?: ActivityFeedQueryParams) =>
    ["activity-feed", params] as const,
}

export function useDiary(params?: DiaryQueryParams) {
  return useQuery({
    queryKey: diaryKeys.list(params),
    queryFn: () => getDiary(params),
  })
}

export function useInfiniteDiary(params?: Omit<DiaryQueryParams, "limit" | "offset">) {
  return useInfiniteQuery({
    queryKey: diaryKeys.infinite(params),
    queryFn: ({ pageParam = 0 }) =>
      getDiary({ ...params, limit: PAGE_SIZE, offset: pageParam }),
    initialPageParam: 0,
    getNextPageParam: (last) => {
      const next = last.offset + last.limit
      return next < last.total_count ? next : undefined
    },
  })
}

export function useActivityFeed(params?: ActivityFeedQueryParams) {
  return useQuery({
    queryKey: diaryKeys.feed(params),
    queryFn: () => getActivityFeed(params),
  })
}

export function useInfiniteActivityFeed(
  params?: Omit<ActivityFeedQueryParams, "limit" | "offset">,
) {
  return useInfiniteQuery({
    queryKey: diaryKeys.feed(params),
    queryFn: ({ pageParam = 0 }) =>
      getActivityFeed({ ...params, limit: PAGE_SIZE, offset: pageParam }),
    initialPageParam: 0,
    getNextPageParam: (last) => {
      const next = last.offset + last.limit
      return next < last.total_count ? next : undefined
    },
  })
}

export function useLogReview() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (data: LogReviewRequest) => logReview(data),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: diaryKeys.all })
      qc.invalidateQueries({ queryKey: ["activity-feed"] })
    },
  })
}

export function useEditReview() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: ({ id, data }: { id: string; data: EditReviewRequest }) =>
      editReview(id, data),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: diaryKeys.all })
      qc.invalidateQueries({ queryKey: ["activity-feed"] })
    },
  })
}

export function useDeleteReview() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (id: string) => deleteReview(id),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: diaryKeys.all })
      qc.invalidateQueries({ queryKey: ["activity-feed"] })
    },
  })
}
